lib/Mojo/WebSocketProxy/Backend/JobAsync.pm

```diff
@@ -137,15 +153,24 @@ sub call_rpc {
             my $api_response;
 
             if ($f->is_done) {
-                my $result = MojoX::JSON::RPC::Client::ReturnObject->new(rpc_response => decode_json_utf8($f->get));
-
-                $_->($c, $req_storage, $result) for @$after_got_rpc_response_hook;
-
-                $api_response = $rpc_response_cb->($result->result);
-                stats_inc("rpc_queue.client.jobs.success", {tags => ["rpc:" . $req_storage->{name}, 'clientID:' . $self->client->id]});
+                try {
+                    my $result = MojoX::JSON::RPC::Client::ReturnObject->new(rpc_response => decode_json_utf8($f->get));
+
+                    $_->($c, $req_storage, $result) for @$after_got_rpc_response_hook;
+
+                    $api_response = $rpc_response_cb->($result->result);
+                    stats_inc("rpc_queue.client.jobs.success", {tags => ["rpc:" . $req_storage->{name}, 'clientID:' . $self->client->id]});
+                }
+                catch {
+                    $log->warnf("Failed to process response of method %s: %s", $method, $_);
+                    stats_inc("rpc_queue.client.jobs.fail",
+                        {tags => ["rpc:" . $req_storage->{name}, 'clientID:' . $self->client->id, 'error:' . $_]});
+                    $api_response = $c->wsp_error($msg_type, 'WrongResponse', 'Sorry, an error occurred while processing your request.');
+                };
             } else {
-                my ($failure) = $f->failure;
-                $log->warnf("method %s failed: %s", $method, $failure);
+                my $failure = $f->is_failed ? $f->failure // '' : 'Rpc request was cancelled. Subscription is not ready.';
```

Comment:
```suggestion
                my $failure = $f->is_failed ? $f->failure // '' : 'RPC request was cancelled. Subscription is not ready.';
```

although I'd suggest splitting these out into separate done/failed/cancelled mappings, rather than putting it all under `->on_ready`

Reply:
@tom-binary I agree that splitting into done/failed/cancel is better, but doing so the lines of code shared among them should be duplicated which doesn't seem reasonable to me.